src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -32,6 +39,46 @@ public synchronized Product getProduct(Long id) {
                 .collect(Collectors.toList());
     }
 
+    @Override
+    public synchronized List<Product> findProducts(String query) {
+        String[] words = query
+                .trim()
+                .toLowerCase()
+                .split(" ");
+
+        return products.stream()
+                .filter(product -> product.getPrice() != null && product.getStock() > 0)
+                .filter(product -> Arrays.stream(words).anyMatch(x -> product.getDescription().toLowerCase().contains(x)))
+                .collect(Collectors.toList());
+
+//        return products.stream()
+//                .filter(product -> product.getPrice() != null && product.getStock() > 0)
+//                .collect(Collectors.toMap(Function.identity(), product -> Arrays.stream(words).filter(x -> product.getDescription().toLowerCase().contains(x)).count()))
+//                .entrySet().stream()
+//                .filter(x -> x.getValue() > 0)
+//                .sorted(Comparator.comparing(Map.Entry<Product,Long>::getValue))
+//                .map(Map.Entry::getKey)
+//                .collect(Collectors.toList());
+    }
+
+    @Override
+    public synchronized List<Product> findProducts(String query, String order, String sortBy) {
```

Comment:
Pass enumtype variable to dao instead of string, parse the string at servlet layer